File: Data Structures & Algorithms/balanced-binary-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        def isbalance(root):
            if not root:
                return [True, 0]

            left = isbalance(root.left)
            right = isbalance(root.right)

            if left[0] and right[0] and abs(left[1] - right[1]) <= 1: # remember to take [1]
                return [True,  max(left[1], right[1]) + 1]
            else:
                return [False,  max(left[1], right[1]) + 1] # here max + 1 

        return isbalance(root)[0]
        

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-7.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {

        if(subRoot == null)
            return true;

        if(root == null)
            return false;

        if(isSameTree(root, subRoot)) // current root pe check karo
            return true; // if it return true good
        else
        {
            return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);

        }
        // tricky part
        // but if is doesnt root ke left and right part se compare 
        
    }

    public boolean isSameTree(TreeNode root, TreeNode subRoot)
    {
        if(root == null && subRoot == null)
            return true;

        if(root == null || subRoot == null)
            return false;

        if(root.val != subRoot.val)
            return false;
            
        return isSameTree(root.left, subRoot.left) && isSameTree(root.right, subRoot.right);
    }

}
